### PR TITLE
Bpk 1757 native panel android

### PR DESCRIPTION
**Pull request by amolgupta**
Native panel implementation and updated demo app

### CODE REVIEW

File: native/android/app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistery.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package net.skyscanner.backpack.demo.data;
```
Comment (shaundon) [inline native/android/app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistery.java]: I think this filename is misspelled - should be `ComponentRegistry.java`.

File: native/packages/react-native-backpack/libraries/Panel/android/gradlew.bat [hunk 2]
```diff
@@ -0,0 +1,84 @@
+@if "%DEBUG%" == "" @echo off
```
Comment (matteo-hertel) [inline native/packages/react-native-backpack/libraries/Panel/android/gradlew.bat]: are we still keeping this? we don't support windows for developments, we can safely gitignore this

File: native/packages/react-native-backpack/stories.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ import { BpkPanel } from './index';
 storiesOf('react-native-backpack/BPKPanel', module)
   .addDecorator(CenterDecorator)
   .add('docs:default', () => (
-    <BpkPanel>
+    <BpkPanel padded>
```
Comment (matteo-hertel) [inline native/packages/react-native-backpack/stories.js]: can we also add the non-padded version as a separtate story?

File: native/packages/react-native-backpack/libraries/Panel/index.android.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* @flow */
+import React, { Component, type Node } from 'react';
+import PropTypes from 'prop-types';
+import { requireNativeComponent, View } from 'react-native';
+
+type Props = {
```
Comment (matteo-hertel) [inline native/packages/react-native-backpack/libraries/Panel/index.android.js]: type needs exporting, see the RN component https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-panel/index.js

File: native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        this.layoutParams = params
+    }
+
+
+    private fun initialize(context: Context, attrs: AttributeSet?, defStyleAttr: Int) {
+
+        val a = context.obtainStyledAttributes(attrs, R.styleable.panel, R.attr.bpk_panel_padding, 0)
+        if (a.hasValue(R.styleable.panel_bpk_panel_padding)) {
+            //default value of padding is true
+            mPadding = a.getBoolean(R.styleable.panel_bpk_panel_padding, true)
```
Comment (matteo-hertel) [inline native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt]: are we confident this is the naming convention we'll use going forward?

yesterday we mentioned we could use just`bpk_padding`

File: native/packages/react-native-backpack/libraries/Panel/android/gradlew.bat (hunk 2 above)
Comment (amolgupta) [reply native/packages/react-native-backpack/libraries/Panel/android/gradlew.bat]: removed

File: native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        val padding = resources.getDimensionPixelOffset(R.dimen.bpkSpacingSm)
+        if (mPadding)
+            this.setPadding(padding, padding, padding, padding)
+        val params = LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT)
+        params.setMargins(padding, padding, padding, padding)
+        this.layoutParams = params
+    }
+
+
+    private fun initialize(context: Context, attrs: AttributeSet?, defStyleAttr: Int) {
```
Comment (tiagoengel) [inline native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt]: seems like you not using `defStyleAttr` here

File: native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+class BpkPanel : LinearLayout {
+
+    private var mPadding: Boolean = false
+
+    constructor(context: Context) : super(context) {
+        this.orientation = LinearLayout.VERTICAL
+    }
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, R.style.Bpk_panel) {
```
Comment (tiagoengel) [inline native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt]: You're receiving `defStyleAttr` in the constructor but passing `R.style.Bpk_panel` to the super class

File: native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package net.skyscanner.backpack
+
+import android.content.Context
+import android.util.AttributeSet
+import android.widget.LinearLayout
+
+
+class BpkPanel : LinearLayout {
```
Comment (tiagoengel) [inline native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt]: you can simplify this with:

```
class BpkPanel(
  context: Context, 
  attrs: AttributeSet?, 
  defStyleAttr: Int = R.style.Bpk_pane
): LinearLayout(context, attrs, defStyleAttr) {

   constructor(context: Context) : this(context, null)
   constructor(context: Context, attrs: AttributeSet?) : this(context, attrs)

   init {
      this.orientation = LinearLayout.VERTICAL
      initialize(context, attrs, defStyleAttr)
   }
```

File: native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    private var mPadding: Boolean = false
+
+    constructor(context: Context) : super(context) {
+        this.orientation = LinearLayout.VERTICAL
```
Comment (tiagoengel) [inline native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt]: Shouldn't this be in all constructors?

File: native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    fun setmPadding(mPadding: Boolean) {
+        this.mPadding = mPadding
+        draw()
+    }
+
+    private fun draw() {
+        this.background = context.getDrawable(R.drawable.border)
+        val padding = resources.getDimensionPixelOffset(R.dimen.bpkSpacingSm)
+        if (mPadding)
```
Comment (tiagoengel) [inline native/packages/react-native-backpack/libraries/Panel/android/native/src/main/java/net/skyscanner/backpack/BpkPanel.kt]: nit: use braces

File: native/packages/react-native-backpack/libraries/Panel/index.android.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+export type Props = {
+  children: Node,
+  padded: boolean,
+};
+/* eslint-disable react/prefer-stateless-function */
+class BpkPanel extends Component<Props, {}> {
+  static propTypes = {
+    padded: PropTypes.bool,
+    ...View.propTypes, // include the default view properties
```
Comment (tiagoengel) [inline native/packages/react-native-backpack/libraries/Panel/index.android.js]: Question about this, do we use this in the native side? Can we add a test for this?

Comment (amolgupta) [reply native/packages/react-native-backpack/libraries/Panel/index.android.js]: Yes, this what is used to lookup the native view params. Will add snapshot tests for all the views once we are sure we are using this approach for native components

**Pull request merged**